data/pigui/modules/system-view-ui.lua

```diff
@@ -408,6 +440,102 @@ function Windows.unexplored.Show()
 	ui.text(lc.UNEXPLORED_SYSTEM_NO_SYSTEM_VIEW)
 end
 
+local function checkSurfPorts(sbody)
+	local children = sbody.children
+	if #children == 0 then return nil end
+	local ports = {}
+	local has_indicators
+	local navTarget = Game.player:GetNavTarget()
+	local selected = systemView:GetSelectedObject().ref
+	for _, child in ipairs(children) do
+		if child.type == "STARPORT_SURFACE" then
+			table.insert(ports, child)
+			local body = child.physicsBody
+			has_indicators = has_indicators or body and body == navTarget or child == selected
+		end
+	end
+	if #ports == 0 then return nil end
+	return ports, has_indicators
+end
+
+local function displaySurfPorts(center, radius, ports, mouseover)
+	local nports = #ports
+	if nports == 0 then return nil end
+	local icon_size = bodyIconSize.x
+	local two_pi = math.pi * 2
+	-- pretty rough estimate
+	local min_rad = icon_size * nports / two_pi
+	radius = math.max(icon_size, min_rad, radius - icon_size / 2)
+
+	local start_ang = 225 * two_pi / 360
+	local step_ang = two_pi / nports
+	local hover_i = 0
+	local hover_center
+	local select_i = 0
+	local select_center
+	for i = 1, nports do
+		local ang = start_ang + step_ang * (i - 1)
+		local p_center = center + Vector2(math.cos(ang), math.sin(ang)) * radius
+		-- HACK we pretend to be a group, to use drawGroupIcons
+		local body = ports[i].physicsBody
+		local group = { hasNavTarget = body and body == Game.player:GetNavTarget() }
+		local isSelected = ports[i] == systemView:GetSelectedObject().ref
+		drawGroupIcons(p_center, _getBodyIcon(ports[i], true), svColor.SYSTEMBODY_ICON, bodyIconSize, group, isSelected)
+		-- hover mouse
+		if (ui.getMousePos() - p_center):length() < icon_size / 2 then
+			hover_i = i
+			hover_center = p_center
+		end
+		if isSelected then
+			select_i = i
+			select_center = p_center
+		end
+	end
+	if hover_i ~= 0 then
+		return ports[hover_i], hover_center
+	elseif select_i ~= 0 and not mouseover then
+		return ports[select_i], select_center
+	else
+		return nil
+	end
+end
+
+local popup = {}
+
+function popup.make()
+	if not popup.enabled then return end
+	ui.popup("system-view-ui-popup", function()
```

Comment:
Since you're hardcoding the popup ID here, instead of calling `popup.enabled = false` later down the line I'd recommend calling `ui.closeCurrentPopup` directly - the draw function will not be called if the popup itself is not open.

Reply:
Indeed! For some reason I thought that the function passed to `ui.popup` would be executed every frame in `popup.make()`, but of course this does not happen, so you don't even need to call anything, the popup itself closes when you click in it.